### PR TITLE
Fix missing poster fetch & add torrent collection search engine enable/disable feature

### PR DESCRIPTION
**Pull request by kiriles90**
* Fix fetching of missing posters in movie_detail view

* Add the ability to enable/disable search engines in the Torrent Collection
(`torrent_collection.js` looks heavily edited but after https://github.com/popcorn-official/popcorn-desktop/pull/1583/files#diff-f7847d725e4a233b43d20b518bc499ceR119 is basically some extra indent and rearranging, the total added code is far less than what it looks like in github's diff)

### CODE REVIEW

File: src/app/lib/views/torrent_collection.js [hunk 1]
```diff
@@ -56,6 +61,32 @@
             });
         },
 
+        togglesengines: function () {
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:64]: can you make these names camelCase?

File: src/app/lib/views/torrent_collection.js [hunk 2]
```diff
@@ -76,179 +107,193 @@
                 return;
             }
 
+            if($('.search_in').is(':visible')) {
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:110]: `.togglesengines` and `.search_in` are used a lot, as well as the icons for them. there's a lot of room for typos here. can we make functions to get these components or constants for their selectors/class names?

File: src/app/lib/views/torrent_collection.js [hunk 3]
```diff
@@ -76,179 +107,193 @@
...
-                        console.debug('1337x search: %s results', data.torrents.length);
-                        var indx = 1, totl = data.length;
-                        data.torrents.forEach(function (item) {
-                            leet.info('https://1337x.to' + item.href).then(function (ldata) {
+            var piratebay = function () {
+                if (Settings.enablethepiratebay) {
+                    return new Promise(function (resolve) {
+                        var results = [];
+                        setTimeout(function () {
+                            resolve(results);
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:128]: why is this here? is this supposed to be a 6 second timeout on searches? if so, please add a comment. if not, you should be calling resolve once you have all the results. we're in node v12, so we can use async functions as well

File: src/app/lib/views/torrent_collection.js [hunk 4]
```diff
@@ -76,179 +107,193 @@
...
+                                    magnet: item.magnetLink,
+                                    seeds: item.seeders,
+                                    peers: item.leechers,
+                                    size: item.size,
                                     index: index
                                 };
-                                indx++;
-                                if (item.title.match(/trailer/i) !== null && input.match(/trailer/i) === null) {
+
+                                if (item.name.match(/trailer/i) !== null && item.name.match(/trailer/i) === null) {
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:152]: how can this condition ever be true? why is it even being checked? please add a comment as to why trailers are not returned

File: src/app/lib/views/torrent_collection.js [hunk 5]
```diff
@@ -76,179 +107,193 @@
...
                                 }
                                 results.push(itemModel);
                                 index++;
-                            }).catch(function (err) {
-                                throw 'nope';
                             });
+                            resolve(results);
+                        }).catch(function (err) {
+                            console.error('tpb search:', err);
+                            resolve(results);
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:161]: do we really want errors to resolve correctly? shouldn't they reject? also, you are going to run into cases where you're resolving twice, which is illegal in node (due to the timeout resolve)

File: src/app/lib/views/torrent_collection.js [hunk 6]
```diff
@@ -76,179 +107,193 @@
...
-                    }).then(function (data) {
-                        console.debug('OMG search: %s results', data.torrents.length);
-                        var indx = 1, totl = data.length;
-                        data.torrents.forEach(function (item) {
-                            omg.info(item.href).then(function (ldata) {
+            var leetx = function () {
+                if (Settings.enable1337x) {
+                    return new Promise(function (resolve) {
+                        var results = [];
+                        setTimeout(function () {
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:171]: same concerns here. please make a torrent search function if the logic is going to be shared

File: src/app/lib/views/torrent_collection.js [hunk 7]
```diff
@@ -76,179 +107,193 @@
...
+                            resolve(results);
+                        });
+                    });
+                }
+            };
+
+            var rarbg = function (retry) {
+                if (Settings.enablerarbg) {
+                    return new Promise(function (resolve) {
+                        var results1 = [];
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:214]: same comments as above

File: src/app/lib/views/torrent_collection.js [hunk 8]
```diff
@@ -76,179 +107,193 @@
...
+                        }).catch(function (err) {
+                            console.error('1337x search:', err);
+                            resolve(results);
+                        });
+                    });
+                }
+            };
+
+            var rarbg = function (retry) {
+                if (Settings.enablerarbg) {
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js]: Please make this name camelCase as well as the other settings. The settings might be better named `enable[search engine name]Search` instead of just `enable[search engine name]`, this way it's clearer what they actually do

File: src/app/lib/views/torrent_collection.js [hunk 9]
```diff
@@ -76,179 +107,193 @@
...
+                                    size: Common.fileSize(parseInt(item.size)),
                                     index: index
                                 };
-                                indx++;
+
                                 if (item.title.match(/trailer/i) !== null && input.match(/trailer/i) === null) {
                                     return;
                                 }
-                                results.push(itemModel);
+                                results1.push(itemModel);
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:239]: why is it results1?

File: src/app/lib/views/torrent_collection.js [hunk 10]
```diff
@@ -76,179 +107,193 @@
...
-                                index: index
-                            };
-
-                            if (item.name.match(/trailer/i) !== null && item.name.match(/trailer/i) === null) {
-                                return;
+                            resolve(results1);
+                        }).catch(function (err) {
+                            console.error('rarbg search:', err);
+                            if (!retry) {
+                                return resolve(rarbg(true));
```
Comment (ArcticZeroo) [inline src/app/lib/views/torrent_collection.js:246]: why is rargb the only one with retries?

File: src/app/lib/views/movie_detail.js [hunk 11]
```diff
@@ -47,7 +47,7 @@
 
       curSynopsis = {old: '', crew: '', cast: '', allcast: '', vstatus: null};
 
-      if (!this.model.get('synopsis') || !this.model.get('rating') || this.model.get('rating') == '0' || this.model.get('rating') == '0.0' || !this.model.get('runtime') || this.model.get('runtime') == '0' || !this.model.get('trailer') || !this.model.get('cover') || this.model.get('cover') == 'images/posterholder.png' || !this.model.get('backdrop') || this.model.get('backdrop') == 'images/posterholder.png' || (Settings.translateSynopsis && Settings.language != 'en')) {
+      if (!this.model.get('synopsis') || !this.model.get('rating') || this.model.get('rating') == '0' || this.model.get('rating') == '0.0' || !this.model.get('runtime') || this.model.get('runtime') == '0' || !this.model.get('trailer') || !this.model.get('poster') || this.model.get('poster') == 'images/posterholder.png' || !this.model.get('backdrop') || this.model.get('backdrop') == 'images/posterholder.png' || (Settings.translateSynopsis && Settings.language != 'en')) {
```
Comment (ArcticZeroo) [inline src/app/lib/views/movie_detail.js:49]: could you add a comment explaining why we are checking these seemingly arbitrary values?

**Review by ArcticZeroo — CHANGES_REQUESTED** [review]
It looks like it will mostly run fine aside from the timeout on resolve and the double resolve, but it would be great if you could make the other requested changes to keep increasing code quality as we add new code

File: src/app/lib/views/torrent_collection.js (hunk 1 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:64]: > can you make these names camelCase?

Yep sure.

File: src/app/lib/views/torrent_collection.js (hunk 3 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:128]: That wasnt me. It was pre-existing code.

File: src/app/lib/views/torrent_collection.js (hunk 4 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:152]: Same for this one, pre-existing code.

File: src/app/lib/views/torrent_collection.js (hunk 5 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:161]: And for this...

File: src/app/lib/views/torrent_collection.js (hunk 7 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:214]: I feel like I'm repeating my self too :D This code isnt mine. I didnt re-write the torrent_collection. I just added a way to disable/enable the search engines.

File: src/app/lib/views/torrent_collection.js (hunk 3 above)
Comment (ArcticZeroo) [reply src/app/lib/views/torrent_collection.js:128]: Oh, I wasn't aware of that. Where is this code from then? We should definitely rewrite it at some point, but if the code already existed I guess it's fine as-is 🤷‍♀️

File: src/app/lib/views/torrent_collection.js (hunk 9 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:239]: Not me!!! :D  

I see what happened.. (though I did write a note about it). These arent changes. I didnt change anything in those function. I just added an if statement for each of them, so I had to also indent. I also re-arranged them so they are tpb, 1337x, rarbg,omgtorrents and have the same order as everywhere else. Other than that all the code they contain is pre-existing code. Github just crapped itselfs a bit trying to figure out the edits. Sorry :)

File: src/app/lib/views/torrent_collection.js (hunk 10 above)
Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:246]: Again.. not my code but probably because rarbg also requires a token to search and also has a 1 request / 2seconds limit. I guess the retry helps?

File: src/app/lib/views/movie_detail.js (hunk 11 above)
Comment (kiriles90) [reply src/app/lib/views/movie_detail.js:50]: Ahh, this is mine :D Checking for missing metadata and fetching them? (could also be added in the movie_list etc but will create a lot of calls to TMDb.

File: src/app/lib/views/torrent_collection.js (hunk 9 above)
Comment (ArcticZeroo) [reply src/app/lib/views/torrent_collection.js:239]: oh, that makes sense. of course I totally missed the note, my bad :) this file should definitely get a rewrite eventually, but that's out of the scope of this PR obviously. for some reason the "resolve conversation" button is missing here

Comment (kiriles90) [reply src/app/lib/views/torrent_collection.js:239]: I do still need to camelCase some names. Also if you have any suggestions renaming some functions or variables to avoid confusion let me know and I'll edit them.

File: src/app/lib/views/movie_detail.js (hunk 11 above)
Comment (ArcticZeroo) [reply src/app/lib/views/movie_detail.js:49]: That makes sense to me, can you add a comment in the code explaining basically that? This way nobody needs to come searching for the PR comments to find out why we're doing this :)

**Comment by kiriles90** [comment]
@ArcticZeroo 
Changed

```
enablethepiratebay -> enableThepiratebaySearch
enable1337x        -> enable1337xSearch
enablerarbg        -> enableRarbgSearch
enableomgtorrent   -> enableOmgtorrentSearch
```

everywhere (`settings.js`, `torrent_collection.tpl`, `torrent_collection.js`)

and the
```
togglethepiratebay -> toggleThepiratebay
togglerarbg        -> toggleRarbg
toggleomgtorrent   -> toggleOmgtorrent
```
functions in `torrent_collection.js`
(`toggle1337x` stayed the same, cant uppercase '1' and nothing else to add to it)

Also skinned the checkboxes (took the ones from the settings page and shrunk them a bit, the styling is the exact same one though)

**Comment by ArcticZeroo** [comment]
Thanks for making those changes! Sorry for making you change a bunch of stuff in code that isn't yours, but improving the existing code even bit by bit is a win! I think this is the only thing left: https://github.com/popcorn-official/popcorn-desktop/pull/1583/files#r463329363 (it's resolved, but you can expand it and see my comment)

**Comment by kiriles90** [comment]
If I understood which one you ment correctly, done!

(also I put the other "//Handle keyboard shortcuts.." note back in its place, seems someone moved it at some point and it was pointing to nothing)

**Comment by kiriles90** [comment]
If I understood which one you ment correctly, done!
(I also moved the "//Handle keyboard shortcuts.." note back in its place, seems it got moved at some point and was in the wrong place)
